### DISCUSSION TITLE
Cannot use realtime subscribe in flutter with package pocketbase

### DISCUSSION THREAD

**Discussion by RaulOn18** [pocketbase/pocketbase#3977] [Q&A]
how to use subsribe realtime message in flutter using pocketbase package, is the below correct? how come there is no error but not realtime still have to reload.

[dart]
```
TextEditingController messagesController = TextEditingController();

  List<RecordModel> allMessages = [];
  bool isFinish = false;
  bool isSended = true;
  bool isDataUser = false;
  String avatarUser = "";
  Future<void> Function()? unsubscribe; // Store the unsubscribe function

  getUserDetail() async {
    final currentUser = widget.pb.authStore.model.id;
    final record = await widget.pb.collection('users').getOne("$currentUser");
    setState(() {
      avatarUser = record.data['avatar'];
      isDataUser = true;
    });
    print("Hasil: ${record.data['name']}");
  }

  getAllMessages() async {
    final records = await widget.pb.collection('messages').getList(
          page: 1,
          perPage: 50,
          sort: "-created",
          expand: "sender_user_id",
        );
    allMessages = records.items;
    setState(() {
      isFinish = true;
    });
    print("recrodAll ${records}");
  }

  void deleteMessage(String id) async {
    await widget.pb.collection('messages').delete(id);
  }

  void insertMessage({senderId, message, timestamp}) async {
    setState(() {
      isSended = false;
    });
    final body = <String, dynamic>{
      "sender_user_id": senderId,
      "message": message,
      "timestamp": DateTime.now().toString(),
    };

    try {
      await widget.pb
          .collection('messages')
          .create(body: body)
          .then(
            (value) => {
              print("inserted"),
              messagesController.text = "",
              getAllMessages()
            },
          )
          .then((value) {
        setState(() {
          isSended = true;
        });
      });
    } catch (error) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text(
            error.toString(),
          ),
        ),
      );
    }
  }

  void logout() async {
    widget.pb.authStore.clear();

    final prefs = await SharedPreferences.getInstance();

    await prefs.remove('pb_auth');

    Navigator.pushReplacementNamed(context, '/');
  }

  @override
  void initState() {
    super.initState();

    getAllMessages();
    getUserDetail();

    widget.pb.collection('messages').subscribe(
      '*',
      (event) async {
        if (event.action == 'create') {
          print("object created realtime");
          allMessages = [...allMessages, event.record!];
          isFinish = true;
        } else if (event.action == 'delete') {
          allMessages.removeWhere((element) => element.id == event.record!.id);
        }
      },
    ).then((unsub) {
      unsubscribe = unsub; // Set the unsubscribe function
    });
  }

  @override
  void dispose() {
    if (unsubscribe != null) {
      unsubscribe!(); // Call the unsubscribe function
    }
    allMessages.clear();
    super.dispose();
  }
```

**Comment by ganigeorgiev** [comment]
I'm not able to reproduce it with the Dart cli and the example from https://github.com/pocketbase/dart-sdk/blob/master/example/example.dart.

Note that if your Dart application is running in the browser or you are using Flutter Web, then this is expected to not work because streamed responses are currently not supported on the web by `dart-lang/http` at the moment.
For this and other limitations, including a workaround with external `fetch_client`, you can check https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#limitations

**Reply by RaulOn18** [reply]
thank you for the answer, but in this case I didn't use Flutter Web, I used flutter native android, for the code there was no error, but for real time it still didn't work properly, it was still necessary to reload the application, then I tried using svelte which is more or less the same code, but I put the subscribe variable `onmount`, which is in flutter `initstate`, am I using it wrong?

**Reply by ganigeorgiev** [reply]
I'm having difficulties understanding your comment and what actually is the issue.

1. Do you mean that ` widget.pb.collection('messages').getList()` returns the expected results, but `widget.pb.collection('messages').subscribe()` doesn't trigger on changes?

2. By saying _"I tried using svelte..."_, do you mean that with the Svelte code the `subscribe` callback get triggered?

I'm not sure that I'll be able to help you debug it as there too many things that could affect it.
It is also not clear from above when the authentication happens and how `widget.pb` is initialized. 
Try to add `catch` block to the subscribe call to see if there is any error.
If not, double check your `messages` List API rule.

**Reply by ganigeorgiev** [reply]
To confirm the `subscribe` call is actually reaching the server you can enable the debug logs by going in the _Admin UI > Logs > Logs settings > Min log level_ and set it to `-4` and then filter with `level = -4` in the searchbar to see the realtime log messages.
If you use PocketBase v0.20.1 you can also use the `--dev` flag to explore all log messages in the terminal.
